Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp

```diff
@@ -308,87 +359,91 @@ std::string Joystick::RampEffect::GetName() const
+bool Joystick::ConstantEffect::UpdateParameters(s16 value)
 {
-  m_effect.type = SDL_HAPTIC_CONSTANT;
-  m_effect.constant.length = RUMBLE_LENGTH_MAX;
-  m_effect.constant.level = (Sint16)(state * 0x7FFF);
-}
+  s16& level = m_effect.constant.level;
+  const s16 old_level = level;
 
-void Joystick::RampEffect::SetSDLHapticEffect(ControlState state)
-{
-  m_effect.type = SDL_HAPTIC_RAMP;
-  m_effect.ramp.length = RUMBLE_LENGTH_MAX;
-  m_effect.ramp.start = (Sint16)(state * 0x7FFF);
+  level = value;
+
+  m_effect.type = level ? SDL_HAPTIC_CONSTANT : DISABLED_EFFECT_TYPE;
+  return level != old_level;
 }
 
-void Joystick::SineEffect::SetSDLHapticEffect(ControlState state)
+bool Joystick::RampEffect::UpdateParameters(s16 value)
 {
-  m_effect.type = SDL_HAPTIC_SINE;
-  m_effect.periodic.period = RUMBLE_PERIOD;
-  m_effect.periodic.magnitude = (Sint16)(state * 0x7FFF);
-  m_effect.periodic.offset = 0;
-  m_effect.periodic.phase = 18000;
-  m_effect.periodic.length = RUMBLE_LENGTH_MAX;
-  m_effect.periodic.delay = 0;
-  m_effect.periodic.attack_length = 0;
+  s16& level = m_effect.ramp.start;
+  const s16 old_level = level;
+
+  level = value;
+  // FYI: Setting end to same as start is odd,
+  // but so is using Ramp effects for rumble simulation.
+  m_effect.ramp.end = level;
+
+  m_effect.type = level ? SDL_HAPTIC_RAMP : DISABLED_EFFECT_TYPE;
+  return level != old_level;
 }
 
-void Joystick::TriangleEffect::SetSDLHapticEffect(ControlState state)
+bool Joystick::PeriodicEffect::UpdateParameters(s16 value)
 {
-  m_effect.type = SDL_HAPTIC_TRIANGLE;
-  m_effect.periodic.period = RUMBLE_PERIOD;
-  m_effect.periodic.magnitude = (Sint16)(state * 0x7FFF);
-  m_effect.periodic.offset = 0;
-  m_effect.periodic.phase = 18000;
-  m_effect.periodic.length = RUMBLE_LENGTH_MAX;
-  m_effect.periodic.delay = 0;
-  m_effect.periodic.attack_length = 0;
+  s16& level = m_effect.periodic.magnitude;
+  const s16 old_level = level;
+
+  level = value;
+
+  m_effect.type = level ? m_waveform : DISABLED_EFFECT_TYPE;
+  return level != old_level;
 }
 
-void Joystick::LeftRightEffect::SetSDLHapticEffect(ControlState state)
+bool Joystick::LeftRightEffect::UpdateParameters(s16 value)
 {
-  m_effect.type = SDL_HAPTIC_LEFTRIGHT;
-  m_effect.leftright.length = RUMBLE_LENGTH_MAX;
-  // max ranges tuned to 'feel' similar in magnitude to triangle/sine on xbox360 controller
-  m_effect.leftright.large_magnitude = (Uint16)(state * 0x4000);
-  m_effect.leftright.small_magnitude = (Uint16)(state * 0xFFFF);
+  u16& level = (Motor::STRONG == m_motor) ? m_effect.leftright.large_magnitude :
+                                            m_effect.leftright.small_magnitude;
+  const u16 old_level = level;
+
+  level = value;
+
+  m_effect.type = level ? SDL_HAPTIC_LEFTRIGHT : DISABLED_EFFECT_TYPE;
+  return level != old_level;
 }
 #endif
 
 void Joystick::UpdateInput()
 {
-  // each joystick is doin this, o well
+  // Each joystick is doing this.. o well..
```

Comment:
🤔 
...this wasn't really the most useful comment before, and the change doesn't make it that much more useful either :/